File: arduino/nodeMCU/BedroomAC.cpp
#include "BedroomAC.h"
#include "Log.h"

BedroomAC::BedroomAC() : _ac(IR_LED_PIN) {
    _state.power       = false;   // 开关：true=开, false=关
    _state.mode        = "cool";  // 模式："cool"制冷 | "heat"制热 | "dry"除湿 | "fan"送风
    _state.temperature = 26;      // 设定温度：16~30 °C
    _state.swing       = false;   // 摆风：true=开, false=关
    _state.windSpeed   = "auto";  // 风速："auto"自动 | "low"低 | "medium"中 | "high"高
    _state.gentle      = false;   // 舒风模式：true=开, false=关
    _state.light       = true;    // 屏显：true=亮, false=灭
    _state.onTimer     = 0;       // 定时开机（分钟），0=关闭
    _state.offTimer    = 0;       // 定时关机（分钟），0=关闭
}

void BedroomAC::begin() {
    _ac.begin();
    LOG_PRINTLN("[BedroomAC] TCL112AC 协议模块已就绪");
    LOG_PRINTF("[BedroomAC] 初始状态: %s\n", _state.toJson().c_str());
    _notifyState();
}

void BedroomAC::setOnStateChanged(OnStateChanged callback) {
    _onStateChanged = callback;
}

void BedroomAC::_notifyState() {
    if (_onStateChanged) {
        String json = _state.toJson();
        _onStateChanged(json);
    }
}

const char* BedroomAC::_cycleWindSpeed(const String &current) {
    if (current == "auto")   return "low";
    if (current == "low")    return "medium";
    if (current == "medium") return "high";
    return "auto";
}

void BedroomAC::_syncAndSend() {
    _ac.setPower(_state.power);

    if (_state.mode == "cool")      _ac.setMode(kTcl112AcCool);
    else if (_state.mode == "heat") _ac.setMode(kTcl112AcHeat);
    else if (_state.mode == "dry")  _ac.setMode(kTcl112AcDry);
    else if (_state.mode == "fan")  _ac.setMode(kTcl112AcFan);

    _ac.setTemp(_state.temperature);

    if (_state.windSpeed == "auto")        _ac.setFan(kTcl112AcFanAuto);
    else if (_state.windSpeed == "low")    _ac.setFan(kTcl112AcFanLow);
    else if (_state.windSpeed == "medium") _ac.setFan(kTcl112AcFanMed);
    else if (_state.windSpeed == "high")   _ac.setFan(kTcl112AcFanHigh);

    _ac.setSwingVertical(_state.swing);
    _ac.setEcono(_state.gentle);
    _ac.setLight(_state.light);
    _ac.setOnTimer(_state.onTimer);
    _ac.setOffTimer(_state.offTimer);

    _ac.send();
}

bool BedroomAC::handleAction(const String &action, const String &params) {
    if (action == "togglePower")        return togglePower();
    if (action == "increaseTemperature") return increaseTemperature();
    if (action == "decreaseTemperature") return decreaseTemperature();
    if (action == "toggleSwing")        return toggleSwing();
    if (action == "setCoolingMode")     return setCoolingMode();
    if (action == "setHeatingMode")     return setHeatingMode();
    if (action == "setDryMode")         return setDryMode();
    if (action == "setFanMode")         return setFanMode();
    if (action == "toggleWindSpeed")    return toggleWindSpeed();
    if (action == "enableGentleMode")   return enableGentleMode();
    if (action == "toggleLight")        return toggleLight();
    if (action == "setOnTimer")         return setOnTimer(params);
    if (action == "setOffTimer")        return setOffTimer(params);
    if (action == "cancelOnTimer")      return cancelOnTimer();
    if (action == "cancelOffTimer")     return cancelOffTimer();

    LOG_PRINTF("[BedroomAC] 未知指令: %s\n", action.c_str());
    return false;
}

bool BedroomAC::togglePower() {
    _state.power = !_state.power;
    LOG_PRINTF("[BedroomAC] 电源 → %s\n", _state.power ? "开" : "关");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::increaseTemperature() {
    if (_state.temperature < 30) _state.temperature++;
    LOG_PRINTF("[BedroomAC] 温度 → %d°C\n", _state.temperature);
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::decreaseTemperature() {
    if (_state.temperature > 16) _state.temperature--;
    LOG_PRINTF("[BedroomAC] 温度 → %d°C\n", _state.temperature);
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::toggleSwing() {
    _state.swing = !_state.swing;
    LOG_PRINTF("[BedroomAC] 摆风 → %s\n", _state.swing ? "开" : "关");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::setCoolingMode() {
    _state.mode = "cool";
    LOG_PRINTLN("[BedroomAC] 模式 → 制冷");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::setHeatingMode() {
    _state.mode = "heat";
    LOG_PRINTLN("[BedroomAC] 模式 → 制热");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::setDryMode() {
    _state.mode = "dry";
    LOG_PRINTLN("[BedroomAC] 模式 → 除湿");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::setFanMode() {
    _state.mode = "fan";
    LOG_PRINTLN("[BedroomAC] 模式 → 送风");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::toggleWindSpeed() {
    _state.windSpeed = _cycleWindSpeed(_state.windSpeed);
    LOG_PRINTF("[BedroomAC] 风速 → %s\n", _state.windSpeed.c_str());
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::enableGentleMode() {
    _state.gentle = true;
    LOG_PRINTLN("[BedroomAC] 舒风 → 开");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::toggleLight() {
    _state.light = !_state.light;
    LOG_PRINTF("[BedroomAC] 屏显 → %s\n", _state.light ? "亮" : "灭");
    _syncAndSend();
    _notifyState();
    return true;
}

uint16_t BedroomAC::_parseTimerMinutes(const String &params) {
    StaticJsonDocument<JSON_DOC_SIZE_RC> doc;
    DeserializationError err = deserializeJson(doc, params);
    if (err || !doc.containsKey("minutes")) {
        LOG_PRINTF("[BedroomAC] 定时器参数解析失败: %s\n", params.c_str());
        return UINT16_MAX;
    }
    uint16_t mins = doc["minutes"];
    if (mins > 720) mins = 720;  // 最大 12 小时
    return mins;
}

bool BedroomAC::setOnTimer(const String &params) {
    uint16_t mins = _parseTimerMinutes(params);
    if (mins == UINT16_MAX) return false;
    _state.onTimer = mins;
    LOG_PRINTF("[BedroomAC] 定时开机 → %d 分钟\n", mins);
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::setOffTimer(const String &params) {
    uint16_t mins = _parseTimerMinutes(params);
    if (mins == UINT16_MAX) return false;
    _state.offTimer = mins;
    LOG_PRINTF("[BedroomAC] 定时关机 → %d 分钟\n", mins);
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::cancelOnTimer() {
    _state.onTimer = 0;
    LOG_PRINTLN("[BedroomAC] 定时开机 → 取消");
    _syncAndSend();
    _notifyState();
    return true;
}

bool BedroomAC::cancelOffTimer() {
    _state.offTimer = 0;
    LOG_PRINTLN("[BedroomAC] 定时关机 → 取消");
    _syncAndSend();
    _notifyState();
    return true;
}

File: android/MyHome/app/src/main/java/com/yuwjoo/myhome/modules/设备/bedroomAC/BedroomAC.kt
package com.yuwjoo.myhome.modules.设备.bedroomAC

import android.os.Handler
import android.os.Looper
import com.yuwjoo.myhome.modules.mqtt.MqttManager
import com.yuwjoo.myhome.modules.mqtt.MqttTopicCallback
import com.yuwjoo.myhome.modules.udp.UdpManager
import com.yuwjoo.myhome.modules.udp.UdpTopicCallback
import org.json.JSONObject

class BedroomAC private constructor() {

    companion object {
        private val _instance: BedroomAC by lazy { BedroomAC().also { it.init() } }

        fun getInstance(): BedroomAC = _instance
    }

    private val callbacks = mutableListOf<ACStateCallback>()
    private var state = ACState()
    private val handler = Handler(Looper.getMainLooper())
    private var initialized = false

    val currentState: ACState
        get() = state

    private fun init() {
        if (initialized) return
        initialized = true

        // 订阅 MQTT 设备状态主题（retained 消息，QoS 0 即可）
        MqttManager.getInstance().subscribe(
            topic = BedroomACConfig.TOPIC_DEVICE,
            qos = 0,
            callback = object : MqttTopicCallback {
                override fun onMessageArrived(topic: String, payload: String) {
                    applyStateFromJson(payload)
                }
            }
        )

        // 订阅 UDP 设备状态主题
        UdpManager.getInstance().subscribe(
            topic = BedroomACConfig.TOPIC_DEVICE,
            callback = object : UdpTopicCallback {
                override fun onMessageArrived(topic: String, payload: Any?) {
                    applyStateFromPayload(payload)
                }
            }
        )
    }

    fun addCallback(callback: ACStateCallback) {
        if (!callbacks.contains(callback)) {
            callbacks.add(callback)
        }
    }

    fun removeCallback(callback: ACStateCallback) {
        callbacks.remove(callback)
    }

    // ── 基本操作 ──

    fun togglePower()          = sendCommand("togglePower")
    fun increaseTemperature()  = sendCommand("increaseTemperature")
    fun decreaseTemperature()  = sendCommand("decreaseTemperature")
    fun toggleSwing()          = sendCommand("toggleSwing")
    fun setCoolingMode()       = sendCommand("setCoolingMode")
    fun setHeatingMode()       = sendCommand("setHeatingMode")
    fun setDryMode()           = sendCommand("setDryMode")
    fun setFanMode()           = sendCommand("setFanMode")
    fun toggleWindSpeed()      = sendCommand("toggleWindSpeed")
    fun enableGentleMode()     = sendCommand("enableGentleMode")
    fun toggleLight()          = sendCommand("toggleLight")

    // ── 定时操作 ──

    fun setOnTimer(minutes: Int) {
        val params = JSONObject().apply { put("minutes", minutes.coerceIn(0, 720)) }
        sendCommand("setOnTimer", params)
    }

    fun setOffTimer(minutes: Int) {
        val params = JSONObject().apply { put("minutes", minutes.coerceIn(0, 720)) }
        sendCommand("setOffTimer", params)
    }

    fun cancelOnTimer()  = sendCommand("cancelOnTimer")
    fun cancelOffTimer() = sendCommand("cancelOffTimer")

    // ── 消息收发 ──

    private fun sendCommand(action: String, params: JSONObject? = null) {
        // MQTT payload 格式：{"action":"togglePower","params":{...}}
        // UDP 需包装为：{"topic":"YHHome/RC/bedroomAC","data":{"action":"...","params":{...}}}
        val actionJson = JSONObject().apply {
            put("action", action)
            if (params != null) put("params", params)
        }

        val acDevice = UdpManager.getInstance().deviceList
            .find { it.deviceType == BedroomACConfig.DEVICE_TYPE }

        if (acDevice != null) {
            UdpManager.getInstance().publish(
                topic = BedroomACConfig.TOPIC_RC,
                payload = actionJson,
                targetIp = acDevice.ipAddress,
            )
        } else {
            MqttManager.getInstance().publish(
                topic = BedroomACConfig.TOPIC_RC,
                payload = actionJson.toString(),
                qos = 1,
            )
        }
    }

    // ── 状态解析 ──

    /**
     * 解析 MQTT 收到的 retained 状态 JSON
     * 格式：{"power":true,"mode":"cool","temperature":26,...}
     */
    private fun applyStateFromJson(jsonStr: String) {
        try {
            applyStateFromPayload(JSONObject(jsonStr))
        } catch (_: Exception) { }
    }

    /**
     * 解析 UDP 收到的状态 payload
     * UDP 格式：{"topic":"...","data":{"power":true,...}}
     * 需要先提取 data 字段
     */
    private fun applyStateFromPayload(payload: Any?) {
        val json = when (payload) {
            is JSONObject -> {
                // UDP 包装格式：有 data 字段则解包
                if (payload.has("data")) payload.getJSONObject("data") else payload
            }
            is String -> try { JSONObject(payload) } catch (_: Exception) { null }
            else -> null
        } ?: return

        val newState = ACState(
            power       = json.optBoolean("power", state.power),
            temperature = json.optInt("temperature", state.temperature),
            mode        = json.optString("mode", state.mode),
            swing       = json.optBoolean("swing", state.swing),
            windSpeed   = json.optString("windSpeed", state.windSpeed),
            gentle      = json.optBoolean("gentle", state.gentle),
            light       = json.optBoolean("light", state.light),
            onTimer     = json.optInt("onTimer", state.onTimer),
            offTimer    = json.optInt("offTimer", state.offTimer),
        )

        if (newState != state) {
            state = newState
            handler.post {
                callbacks.forEach { it.onStateChanged(state) }
            }
        }
    }
}
